tensilelite/Tensile/KernelWriter.py

```diff
@@ -3030,21 +3048,27 @@ def kernelBody( self, kernel, tensorParametersA, tensorParametersB ):
       module.add(self._wait(kernel, tensorParameters1st, tensorParameters2nd, 0, -1, -1, "2wait for global read"))
       module.add(self._syncThreads(kernel))
 
+      # init local write offsets to nondtl loads in tail loop.
+      module.add(self.lwaInitAddressesForDTLTailLoop(kernel, tensorParameters1st))
+      module.add(self.lwaInitAddressesForDTLTailLoop(kernel, tensorParameters2nd))
+
       # the following read/write addresses could be modified in recalcLocal(Read|Write)Addresses due to policy change
       self.oriLraA = None # back up original local read address vgpr
       self.oriLraB = None
       self.oriLraM = None
       self.oriLwaA = None # back up original local write address vgpr
       self.oriLwaB = None
       self.oriLwaM = None
-      if not kernel["NoLdsWriteCode"]:
+      if not kernel["NoLdsWriteCode"] or kernel["NonDTLTailLoopA"] or kernel["NonDTLTailLoopB"]:
         # tail: local write
-        module.addComment1("local write a")
-        tempLWCodeModA = self.localWriteDo(kernel, tensorParametersA)
-        module.add(tempLWCodeModA)
-        module.addComment1("local write b")
-        tempLWCodeModB = self.localWriteDo(kernel, tensorParametersB)
-        module.add(tempLWCodeModB)
+        if not (kernel["DirectToLdsA"] and not kernel["NonDTLTailLoopA"]):
```

Comment:
I think this if statement is also redundant.
You put if condition in localWriteDo as well.
We need only one of them.